Fighters

DmitryIldyukov opened the pull request.

DmitryIldyukov committed:
Реализовал MVP для fighters

DmitryIldyukov committed:
Добавил шанс критической атаки и множитель атаки

DmitryIldyukov committed:
Добавил проверку на минимальное кол-во бойцов

DmitryIldyukov committed:
Поправил названия некоторых переменных и методов

Nafanyan commented:
Сделал я таких персонажей)
<img width="267" alt="image" src="https://github.com/user-attachments/assets/7609a05b-6c15-4b76-8874-c70cab1c60ab">

И у меня возник вопрос как атакующий наносит 0 урона, а защищающийся получает 1. Я бы понял если бы было наоборот, нанёс 1, а прошло 0 из-за брони, но тут как будто как-то странно)
<img width="381" alt="image" src="https://github.com/user-attachments/assets/d3bf73cd-8530-4685-931c-edceab64328c">

Nafanyan commented on Gladiators/Models/Fighters/Fighter.cs:
```diff
@@ -0,0 +1,96 @@
+    public int TakeDamage( int opponentDamage )
+    {
+        int startHealth = CurrentHealth;
+        int newHealth = CurrentHealth - ( Math.Max( opponentDamage - ArmorPoints, 1 ) );
```
если оппонент нанесет 0 урона, то боец получит 1 единицу урона, как так)

DmitryIldyukov replied:
Было сделано, чтобы не было бесконечного цикла при поединке, когда у бойцов очки брони больше, чем урон) 
Решил добавить состояние для брони

Nafanyan commented on Gladiators/Models/Fighters/Fighter.cs:
```diff
@@ -0,0 +1,96 @@
+
+    public bool IsAlive => CurrentHealth > 0;
+
+    public Fighter( string name,
```
```c#
    public Fighter(
        string name,
        IFighterType fighterType,
        IRace race,
        IWeapon weapon,
        IArmor armor )
```

Nafanyan commented on Gladiators/Services/FighterService.cs:
```diff
@@ -0,0 +1,204 @@
+        return armor;
+    }
+
+    private string GetNonEmptyStringFromConsole( string message,
```
``` c#
    private string GetNonEmptyStringFromConsole(
        string message,
        string errorMessage = "Неверный ввод. Введенная строка не может быть пустой. Попробуйте снова." )
```

Nafanyan commented on Gladiators/GameMaster/GameManager.cs:
```diff
@@ -0,0 +1,154 @@
+            return;
+        }
+
+        Console.WriteLine( message );
```
можно 
``` c#
Console.WriteLine( message + '\n');
```
а Console.WriteLine(); убрать)

Nafanyan replied:
аналогично в других местах

Nafanyan commented on Gladiators/GameMaster/GameManager.cs:
```diff
@@ -0,0 +1,154 @@
+            ShowFighterCountError( fighterCount );
+            return;
+        }
+        IFighter winner = null;
```
какой-то монолит кода, давай логические пустые строки поставим)

Nafanyan commented on Gladiators/GameMaster/GameManager.cs:
```diff
@@ -0,0 +1,154 @@
+    {
+        int damage = fighter.CalculateDamage();
+        int damageTaken = opponent.TakeDamage( damage );
+        Console.WriteLine( $"{fighter.Name} атакует {opponent.Name} и наносит ему {damage} урона." );
```
небольшая путанница с "наносит" и "получает", может сделать что-то вроде:
``` c#
        Console.WriteLine( $"{fighter.Name} атакует {opponent.Name} с силой равной {damage}." );
        if ( damageTaken < damage )
        {
            Console.WriteLine( $"{opponent.Name} частично блокирует атаку, получает {damageTaken} урона и {( opponent.IsAlive ? "выживает" : "погибает" )}." );
        }
        else
        {
            Console.WriteLine( $"{opponent.Name} получает {damageTaken} урона и {( opponent.IsAlive ? "выживает" : "погибает" )}." );
        }
```

???

DmitryIldyukov replied:
Да, так более понятно)

Nafanyan reviewed (changes requested):
В целом очень даже хорошо!

Только есть некоторые вопросики)

DmitryIldyukov committed:
Правки

1) По замечаниям
2) Исправил баг, когда attackMultiplier мог быть равен 0
3) Добавил свойство Strength для бойца и показываю его в информации о бойце. Ранее выводился параметр урона, который был не всегда верный из-за множителя урона и критической атаки

DmitryIldyukov committed:
Добавил состояние брони

При получении урона броня изнашивается
Кол-во урона, которое броня может поглотить, высчитывается по формуле: состояние брони * кол-во очков брони, состояние брони уменьшается по формуле: Текущее состояние брони - Max(урон противника * 0.5, 5)
При восстановлении сил состояние брони становится 100%

Nafanyan reviewed (approved):
Всё круто, отлично может squash'шить и мержить)
